### ISSUE TITLE
Add support for fees that include taxes

### ISSUE THREAD

**Issue by felixarntz** [easydigitaldownloads/Easy-Digital-Downloads#4200]
Currently one can only add fees that exclude taxes. The plugin does not support fees which include taxes, it basically overrides the global setting by adding a filter in `edd_get_cart_fee_tax()` for example.

I'm proposing to add support for fees that include taxes, just like we have for regular download products. I'm currently using a pretty hacky solution to (kind of) fix this problem (which I can share if desired), but I would appreciate if EDD actually supported it in a solid way.

Before we approach this issue, we'd have to think about whether we would take the global setting into account for those fees or whether we made it an additional argument for every individual fee itself. I prefer the latter approach since fees can be of any kind and the way that they are handled may differ. Also, if we just took the global setting, we'd break backwards compatibility with people who have the "fees included" setting active but define their fees without tax.

Maybe we could add a new property to a fee called `tax_mode` and then allow values like 'none', 'include' or 'exclude'. 'exclude' would be the default (for compatibility), and we could migrate the `no_tax` property to this new one by setting the `tax_mode` to 'none' for those.
Of course this enhancement would go beyond that in terms of how taxes, subtotal and total are calculated, so we'd need to work on these processes as well.

Opinions on this?

**Comment by pippinsplugins** [comment]
@felixarntz I'm not clear what you mean by we don't support taxes on fees. When `no_tax` is set to `false`, tax gets added to the fees.

**Comment by felixarntz** [comment]
Yes, taxes on fees are supported, but they are always _added_ on top of the fee amount. I'm talking about the possibility to add fees where their taxes are automatically _included_ in the fee amount you specify. Similar like it is possible for regular products as well. Basically adding fee prices inclusive of tax. Right now we can only add fee prices exclusive of tax.

**Comment by pippinsplugins** [comment]
Ah! That makes more sense, thanks :)

**Comment by mintplugins** [comment]
This is causing incorrect tax amounts when using EDD Recurring sign-up fees if you have taxes set to be included in prices.

**Set prices to include tax:**
<img width="831" alt="screen shot 2016-11-11 at 10 35 43 am" src="https://cloud.githubusercontent.com/assets/1638128/20220221/a5e2af2e-a7fa-11e6-92ce-17de7534d407.png">

**Create recurring product with a signup fee:**
<img width="705" alt="screen shot 2016-11-11 at 10 33 44 am" src="https://cloud.githubusercontent.com/assets/1638128/20220239/bb6b55c6-a7fa-11e6-80e7-2ea6b92943f0.png">

**Total should be $200 as taxes are set to be included in the price.**
<img width="864" alt="screen shot 2016-11-11 at 10 34 33 am" src="https://cloud.githubusercontent.com/assets/1638128/20220297/fa931b1c-a7fa-11e6-84f3-f2f2c12e0df7.png">

I believe this is more of a bug than a feature/enhancement. If you have taxes set to be included in the price, taxes are calculated wrong for fees - including sign-up fees in recurring.

**Comment by mintplugins** [comment]
I have a bit of a wild proposal which is likely not something we'll do - but I wanted to throw it out there and see. 

Instead of making fees work with prices where taxes are included in the **entered** price, we replace the option to **enter** prices including tax, and instead make it an option to simply **display** prices including tax.

**The way it currently is (enter prices including tax):**
- Make a product and set the price to $10.
- Set taxes to 10%
- The customer always sees the price as "$10 including tax".
- At checkout, the amount of tax actually charged is $0.81 because the price was actually $9.09.

**My proposed new way (display prices including tax):**
- Make a product and set the price to $9.09.
- Set taxes to 10%
- The customer always sees the price as "$10 including tax".
- At checkout, the amount of tax actually charged is $0.81 because the price was always $9.09.

**What this changes:**
- The way prices are entered has to be changed from $10 to $9.09 (price without tax).
- Internally, taxes are still calculated "on top" of the price" - simple the front-end display is filtered to show "$10 including tax". The internal math functions no longer have to worry about separating out the taxes and should solve this fee issue as nothing is actually different internally with taxation for excluding or including.

**Backwards Compatibility:**
For stores that already use the "Enter prices including tax" option, nothing changes for them. They still have that option. However, if they run into this bug, they can simply switch to use the new "Display Prices including tax" instead. They would have to then modify the entered price of their product. 

**Forward Motion:**
New stores and stores not currently using the "Enter prices including tax" option will not see the option any longer. It will be replaced for them with the "Display prices including tax".

**Legal Questions:**
In my understanding, countries like Germany require that prices always be displayed including tax. This solution still fits with laws requiring that. 

**My thoughts:**
- I believe this is much clearer to the site owner because they are aware that the actual price of the product is $9.09 and tax is charged on top of that. 

- It also would help to simplify the EDD tax calculation functions as we would no longer need to worry about making changes to the "including tax" issues like this one.

**Comment by pippinsplugins** [comment]
@mintplugins So basically put the responsibility of calculating the pre-tax amount on the store owner?

**Comment by mintplugins** [comment]
@pippinsplugins Exactly, yes.

**Comment by pippinsplugins** [comment]
I'm not immediately opposed or in support of it. Let's think on it.

**Comment by mintplugins** [comment]
I proposed this idea to a client having this issue (tax being added "on top" of signup fee despite having taxes set to be included in the price). 

They are based in Germany and don't see any issue with it legally. They also said they wouldn't mind having to adjust their prices:

https://secure.helpscout.net/conversation/275104971/46368/?folderId=749231

**Comment by pippinsplugins** [comment]
That's good news.

**Comment by felixarntz** [comment]
I'm not really convinced of that approach since having to calculate the tax amount is not user-friendly. If it was a lot easier to implement, it might be worth considering, but generally the user should be able to enter final fee amounts and have the included taxes being calculated.

**Comment by mintplugins** [comment]
@felixarntz Just to throw out an additional possibility, what if while you were entering the price, beside the field it showed you what the output would be so you could keep adjusting the price until it was correct. This way you don't need to do any actual math but just pay with the entered price. 

For example, if you entered $9.00 (at 10% tax), beside the field it would show "$9.90 Including Tax". You could then adjust the price until it came out to exactly $10. 

If you had many different tax rates set up, it could show you the price in all locations. For example, if you are charging taxes in both Canada (13%), and in Germany (19%), if you entered $9, beside/beneath the price field it would show:

- Canada: $10.17 Including Tax
- Germany: $10.71 Including Tax

**Comment by felixarntz** [comment]
That would be nicer in terms of UX, but I'm still wary about this. I think the best solution would be to have an option for how to enter fee amounts, similar to how an option exists for regular prices.

There are cases where some wants to have a product price or fee that is always 9.99, regardless of tax, just so that it always has that "psychological advantage" of being a .99 price.
There are other cases where there is one fixed product price or fee amount, and the taxes should just be added.

In short, I think actual products and fees should have similar behaviors. Also, by making the tax handling of fees an option, we would stay backward-compatible simply by setting the default value to "exclude tax".

**Comment by mintplugins** [comment]
You are most likely correct. While it would greatly simplify the code and remove a lot of potential bugs, it may be too late in the game to remove the "price includes tax" functionality.

**Comment by michaelbeil** [comment]
Still pondering on this, but I am not too keen on changing the "price includes tax" functionality from what it is. I would be more apt to give some better descriptions and tooltips. 

Having to require further calculations of a price total would be rather poor UX, IMHO.

**Comment by mintplugins** [comment]
@m I would agree. We could open a separate issue for tools tips and descriptions. I would argue that we not add any new options. 

Rather, we just need to fix the bug happening with the way that taxes are calculated for fees when taxes are included. So to clarify for this issue:

**The Bug:** 
Taxes not calculated correctly on fees when prices are set to "include tax". Rather than calculating the tax to be included in entered amount, it is being added to the entered amount.

**Comment by michaelbeil** [comment]
Yes, taxes not calculating correctly on fees is a bug.

**Comment by mintplugins** [comment]
Customer Ticket:
https://secure.helpscout.net/conversation/434682291/67018/

**Comment by mintplugins** [comment]
Customer Ticket:
https://secure.helpscout.net/conversation/575648679/81694?folderId=749231

**Comment by mintplugins** [comment]
This will require some major testing with anything that uses the Fees API, especially Discounts Pro, and anything that has negative fees.

**Comment by shaynesanderson** [comment]
I wanted to follow up here and see if this is going to get any traction. I have a customer asking about this functionality but I'm unsure on how to reply since this thread is quite old.

**Comment by LisaCee** [comment]
> I wanted to follow up here and see if this is going to get any traction. I have a customer asking about this functionality but I'm unsure on how to reply since this thread is quite old.

I'm not quite sure if this is what you're looking for, but the option to apply taxes to a fee was added for manual orders in 3.0 recently. [comment](https://github.com/easydigitaldownloads/easy-digital-downloads/pull/8581#issuecomment-827792453)

**Comment by ashleyfae** [comment]
@shaynesanderson To be honest it's just low priority, as the impact is quite small. It's not going to be even considered until after 3.0.
